File: Homework/Week3/mergesort.java
import java.util.Arrays;

public class mergesort {
    public static int[] merge (int[] left, int[] right) {
        int i = 0;
        int j = 0;
        int k = 0;

        int[] result = new int[left.length + right.length];

        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                result[k] = left[i];
                k++;
                i++;
            } else {
                result[k] = right[j];
                k++;
                j++;
            }
        }

        while (i < left.length) {
            result[k] = left[i];
            k++;
            i++;
        }

        while (j < right.length) {
            result[k] = right[j];
            k++;
            j++;
        }

//        System.out.println(Arrays.toString(result));

        return result;
    }

    public static int[] mergeSortHelper(int[] arr, int start, int end) {

        if (start == end) {
            int [] result = new int[1];
            result[0] = arr[start];
//            System.out.println(Arrays.toString(result));
            return result;
        }

        int mid = (start + end) /2;
        int[] left = mergeSortHelper(arr, start, mid);
        int[] right = mergeSortHelper(arr, mid + 1, end);

        System.out.println(Arrays.toString(arr));
        return merge(left, right);
    }

    public static int[] mergeSort(int[] arr) {
        arr = mergeSortHelper(arr,0,arr.length-1);
        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 45,2,67,8};
        int[] result = mergeSort(arr);
        System.out.println(Arrays.toString(result));
    }
}

File: Homework/Week3/mergeSort.js
function merge (left, right) {
    let i = 0;
    let j = 0;

    let returnArray = new Array(0);

    while (i < left.length && j < right.length) {
        if(left[i] < right[j]) {
            returnArray.push(left[i]);
            i++;
        } else {
            returnArray.push(right[j]);
            j++;
        }
    }

    while (i < left.length) {
        returnArray.push(left[i]);
        i++;
    }

    while (j < right.length) {
        returnArray.push(right[j]);
        j++;
    }

    return (returnArray);
}

function mergeHelper (arr,start,end) {

    if (start === end) {
        let resultArray = [];
        resultArray.push(arr[start]);
        return resultArray;
    }

    let mid = Math.floor((start + end)/2);
    let left = mergeHelper(arr,start,mid);
    let right = mergeHelper(arr,mid+1,end);

    let returnArray = merge(left,right);

    console.log(returnArray);

    return returnArray;
}

function mergesort(array) {
    let returnArray = mergeHelper(array,0,array.length-1);
    console.log(returnArray);
    return returnArray;
}

mergesort([5,4,3,2,1]);